src/coinsselectionalgorithm.hpp

```diff
@@ -0,0 +1,280 @@
+/*!
+  This class provides a specific implementation of the solving routine.
+  In this implementation, a binary tree is considered as the combination of excluding/including each coin.
+  This would lead to a number of combinations equal to 2^problemDimension with a brute force strategy.
+  The algorithm doesn't rely on simple brute force strategy, instead two additional aspects are taken into account for
+  speeding up the algorithm and avoiding exploring branches which would not give an improved solution (with respect to
+  the temporary optimal one): backtracking and bounding.
+  Starting with an "all coins unselected" setup, the algorithm recursively explores the tree (from biggest coin toward smallest
+  coin) opening two new branches, the first one excluding the current coin, the second one including the current coin; when a
+  leaf is reached, the output variables are checked to identify if an improved solution (with respect to the temporary optimal
+  one) is found and eventually marked as the new temporary optimal solution.
+  The tree actual exploration differs very significantly from the tree full exploration thanks to:
+  +] backtracking (1): given that at a certain recursion, including a new coin would automatically increase both the temporary
+     total amount as well as the temporary total size, if during the tree exploration the two upper-limit constraints associated
+     to target amount plus offset and to total size are broken then all the branches from the current recursion on are cut;
+     this is done in order to avoid reaching leaves that would certainly be not admissible with respect to these two constraints.
+  +] backtracking (2): given that at a certain recursion, the highest total amount reachable is computed as the sum of current
+     total amount and of all the amounts of coins from the current recursion on, if during the tree exploration this sum does not
+     exceed the lower-limit associated to target amount then all the branches from the current recursion on are cut; this is done
+     in order to avoid reaching leaves that would certainly be not admissible with respect to this constraint.
+  +] bounding: given that at a certain recursion, the highest total selection reachable is computed as the sum of current total
+     selection and of the quantity of coins from the current recrusion on, if during tree exploration this sum does not exceed
+     the temporary optimal solution (ties are handled prioritizing low total amount) then all the branches from the current
+     recursion on are cut; this is done in order to avoid reaching leaves that would certainly not improve the temporary optimal
+     solution.
+*/
+class CCoinsSelectionBranchAndBound : public CCoinsSelectionAlgorithm
+{
+protected:
+    // auxiliary
+    //! The array of cumulative amounts (considered summing amounts from index to end of amounts array)
+    const CAmount* cumulativeAmountsForward;
```

Comment:
Again use STL container or `std::unique_ptr<CAmount[]>` so we don't have to remember to do what RAII already does

Reply:
An analogous consideration about `tmpSelection`

Reply:
Moved from raw pointers to vectors.